distributed/service.py

```diff
@@ -0,0 +1,426 @@
+tasks downstream of it can run.
+
+---------------------------------------------------------------------------------------
+
+Dask handles startup/shutdown, resilience, peer discovery, leader election, RPC
+interfaces, and composition for Services:
+
+Startup/shutdown
+----------------
+
+* Once the *first* input key to a service task is ready (state ``memory``), service
+  instances are started on all workers in the cluster.
+* Once the last output key is produced by a service, all instances are told to stop.
+* Any instance can error the service task; all other instances will be told to stop.
+* Any instance can restart the service task; all other instances will be told to stop,
+  then re-launched from scratch.
+* If the service task is cancelled (because keys downstream from it are released, etc.),
+  all instances are told to stop.
+
+Resilience
+----------
+
+* If a worker that's running a service leaves, the entire service task is restarted on
+  all workers. TODO change once `peer_joined`/`peer_left` are added.
+* If keys downstream of a service need to be recomputed, the service will be rerun, just
+  like regular tasks are.
+* The service is informed at runtime which inputs to expect and which outputs to
+  produce, so partial recomputation (some outputs already in memory, some not) can be
+  handled correctly.
+
+Peer discovery and RPC
+----------------------
+
+* At startup, services are given a list of their peers as RPC handles.
+* The RPC handle allows a `Service` to call arbitrary methods on its peers on other
+  workers.
+* Loopback calls (using the RPC to call yourself) take a fastpath, bypassing
+  serialization and the network stack.
+
+Leader election
+---------------
+* At startup, one instance is designated as the "leader", making it easier to coordinate
+  initialization tasks that need to run exactly once.
+
+Composition
+-----------
+
+One Service's outputs can be another Service's inputs.
+
+A `Service` can depend on multiple input Services, or both normal tasks and Services, or
+have no dependencies at all. Note, though, that input keys are always treated as a flat
+structure (`Service.add_key` is called, one key at a time, from whatever worker already
+holds the key). Therefore, if particular inputs need to be co-located, the `Service` is
+responsible for transferring them around internally. (Doing just that in clever ways is
+often the whole point of a `Service`.)
+
+                 o   o   o
+                  \  |  /
+              combiner-service
+          /   /        \   \   \
+         o   o          o   o   o
+          \ /            \  |  /
+     bar-service        baz-service
+     /   |   \          /  /  \  \
+    o    o    o        i  i    i  i
+      \  |   /          \  \  /  /
+     foo-service       normal-tasks
+
+Restrictions
+~~~~~~~~~~~~
+
+If a task depends on a `Service`, that must be its only dependency. For example, this
+will raise an error when the graph is submitted to the scheduler:
```

Comment:
Can you elaborate a bit further why we need this restriction?

I get that a `service` is evaluated as some kind of "apply on everything", similar to a "map_partitions" and we cannot granularly pass dependencies to it, at the very least this may make things much more complex.

I could easily see a version und usages where we allow dependencies being passed to the service but a dependency may be passed to the service but this dependency is passed to every service instance. 

```mermaid
graph BT
O1 --> D1
O2 --> D2
O3 --> D3
K --> D1
K --> D2
K --> D3
foo-service --> O1
foo-service --> O2
foo-service --> O3
I1 --> foo-service
I2 --> foo-service
```